From 5444dbcd690f9e11b78d6e463b0bd90b6b99ca52 Mon Sep 17 00:00:00 2001
From: IlyaSergeyev <ilya.sergeev@intel.com>
Date: Wed, 14 Sep 2022 14:48:52 +0800
Subject: [PATCH 3/4] FROMGIT: iHD: [Encode] VP9 Enc more secure code for
 unaligned height w/a

memcpy replaced with more secure memcpy_s
one more check after surface lock

(cherry picked from commit 3fcfd6681d418d4ea55d687369ccdd0f0693f9d7 https://github.com/intel/media-driver.git)

BUG=b:231639580
TEST= emerge- libva-intel-media-driver

Signed-off-by: Gareth Yu <gareth.yu@intel.com>
---
 .../agnostic/gen12/codec/hal/codechal_vdenc_vp9_g12.cpp  | 9 +++++++--
 1 file changed, 7 insertions(+), 2 deletions(-)

diff --git a/media_driver/agnostic/gen12/codec/hal/codechal_vdenc_vp9_g12.cpp b/media_driver/agnostic/gen12/codec/hal/codechal_vdenc_vp9_g12.cpp
index 104fe224c9b4..3e9bf2293e56 100644
--- a/media_driver/agnostic/gen12/codec/hal/codechal_vdenc_vp9_g12.cpp
+++ b/media_driver/agnostic/gen12/codec/hal/codechal_vdenc_vp9_g12.cpp
@@ -3470,6 +3470,11 @@ void CodechalVdencVp9StateG12::fill_pad_with_value(PMOS_SURFACE psSurface)
 
         uint8_t *src_data   = (uint8_t *)m_osInterface->pfnLockResource(m_osInterface, &(psSurface->OsResource), &lockFlags);
 
+        if (!src_data)
+        {
+            return;
+        }
+
         uint8_t *src_data_y = src_data + psSurface->dwOffset;
 
         uint32_t y_plane_size      = psSurface->dwPitch * psSurface->dwHeight;
@@ -3482,7 +3487,7 @@ void CodechalVdencVp9StateG12::fill_pad_with_value(PMOS_SURFACE psSurface)
 
         if (src_data_y_end > src_data_y_end - y_pad_length)
         {
-            memcpy(src_data_y_end, src_data_y_end - y_pad_length, y_pad_length);
+            memcpy_s(src_data_y_end, y_pad_length, src_data_y_end - y_pad_length, y_pad_length);
         }
 
         uint32_t uv_plane_size      = (psSurface->dwPitch * psSurface->dwHeight)/2;
@@ -3491,7 +3496,7 @@ void CodechalVdencVp9StateG12::fill_pad_with_value(PMOS_SURFACE psSurface)
 
         if (src_data_uv_end - y_pad_length > src_data_y_end)
         {
-            memcpy(src_data_uv_end, src_data_uv_end - y_pad_length, y_pad_length);
+            memcpy_s(src_data_uv_end, y_pad_length, src_data_uv_end - y_pad_length, y_pad_length);
         }    
 
         m_osInterface->pfnUnlockResource(m_osInterface, &(psSurface->OsResource));
-- 
2.31.0

